packages/insured-bridge-relayer/src/Relayer.ts

```diff
@@ -42,66 +47,73 @@ export class Relayer {
+    // out Finalized relays.
+    const relayableDeposits: RelayableDeposits = this.getRelayableDeposits();
     if (Object.keys(relayableDeposits).length == 0) {
       this.logger.debug({ at: "Relayer", message: "No relayable deposits for any whitelisted tokens" });
       return;
     }
+
+    // Fetch pending relays (if any) for each relayable deposit and then decide whether to submit a relay (and what
+    // type of relay) or dispute.
     for (const l1Token of Object.keys(relayableDeposits)) {
       this.logger.debug({
         at: "Relayer",
         message: `Processing ${relayableDeposits[l1Token].length} relayable deposits for L1Token`,
         l1Token,
       });
       for (const relayableDeposit of relayableDeposits[l1Token]) {
-        // If deposit has a pending relay, then validate its relay params (i.e. any data not included in the deposit
-        // hash). If relay params are invalid, then we should skip it so that we don't speed up an invalid relay.
-        // If we cannot find the relay, then we should not skip it because we can still slow/instant relay it.
         const pendingRelay: Relay | undefined = this.l1Client.getRelayForDeposit(l1Token, relayableDeposit.deposit);
         if (pendingRelay) {
-          const isRelayValid = await this.isRelayValid(pendingRelay, relayableDeposit.deposit);
-          if (!isRelayValid.isValid) {
+          // We need to perform some prechecks on the relay before we attempt to submit a relay. First, we need to check
+          // if the relay has expired, for if it has then we cannot do anything with it except settle it. Second, we need
+          // to check the pending relay's parameters (i.e. any data not included in the deposit hash) and verify that
+          // they are correct. If they are not then we have an option to dispute it, or just ignore it as a potential
+          // speedup candidate.
+          const relayStatus = await this.getRelayStatus(pendingRelay, relayableDeposit.deposit);
+          if (relayStatus.relayExpired.isExpired) {
+            this.logger.debug({
+              at: "Relayer",
+              message: "Pending relay has expired, ignoring",
+              pendingRelay,
+              relayableDeposit,
+              expirationTime: relayStatus.relayExpired.expirationTime,
+              contractTime: relayStatus.relayExpired.contractTime,
+            });
+            continue;
+          } else if (relayStatus.relayDisputable.canDispute) {
             this.logger.debug({
               at: "Relayer",
-              message: "Pending relay is invalid, ignoring",
+              message: "Pending relay is invalid",
               pendingRelay,
               relayableDeposit,
-              reason: isRelayValid.reason,
+              reason: relayStatus.relayDisputable.reason,
             });
+            // TODO: Optionally dispute the relay here if DISPUTE_MODE=true
```

Comment:
I think Relay and Dispute are related since we should only speed up VALID relays, and determining whether a relay is valid is the same logic for determining when to dispute.